src/statement.js

```diff
@@ -518,61 +518,73 @@ pp.parseClass = function(node, isStatement) {
+
+  method.kind = "method"
+  method.static = tryContextual("static")
+  let isGenerator = this.eat(tt.star)
+  let isAsync = false
+  let isGetSet = false
+  if (!isGenerator) {
+    if (this.options.ecmaVersion >= 8 && tryContextual("async", true)) {
       isAsync = true
-      this.parsePropertyName(method)
+    } else if (tryContextual("get")) {
+      isGetSet = true
+      method.kind = "get"
+    } else if (tryContextual("set")) {
+      isGetSet = true
+      method.kind = "set"
     }
-    method.kind = "method"
-    let isGetSet = false
-    if (!method.computed) {
-      let {key} = method
-      if (!isGenerator && !isAsync && key.type === "Identifier" && this.type !== tt.parenL && (key.name === "get" || key.name === "set")) {
-        isGetSet = true
-        method.kind = key.name
-        key = this.parsePropertyName(method)
-      }
-      if (!method.computed && !method.static && (key.type === "Identifier" && key.name === "constructor" ||
-          key.type === "Literal" && key.value === "constructor")) {
-        if (hadConstructor) this.raise(key.start, "Duplicate constructor in the same class")
-        if (isGetSet) this.raise(key.start, "Constructor can't have get/set modifier")
-        if (isGenerator) this.raise(key.start, "Constructor can't be a generator")
-        if (isAsync) this.raise(key.start, "Constructor can't be an async method")
-        method.kind = "constructor"
-        hadConstructor = true
-      } else if (method.static && key.type === "Identifier" && key.name === "prototype") {
-        this.raise(key.start, "Classes may not have a static property named prototype")
-      }
-    }
-    this.parseClassMethod(classBody, method, isGenerator, isAsync)
-    if (isGetSet) {
-      let paramCount = method.kind === "get" ? 0 : 1
-      if (method.value.params.length !== paramCount) {
-        let start = method.value.start
-        if (method.kind === "get")
-          this.raiseRecoverable(start, "getter should have no params")
-        else
-          this.raiseRecoverable(start, "setter should have exactly one param")
-      } else {
-        if (method.kind === "set" && method.value.params[0].type === "RestElement")
-          this.raiseRecoverable(method.value.params[0].start, "Setter cannot use rest params")
-      }
+  }
+  if (!method.key) this.parsePropertyName(method)
+  let {key} = method
+  if (!method.computed && !method.static && (key.type === "Identifier" && key.name === "constructor" ||
+      key.type === "Literal" && key.value === "constructor")) {
+    if (isGetSet) this.raise(key.start, "Constructor can't have get/set modifier")
+    if (isGenerator) this.raise(key.start, "Constructor can't be a generator")
+    if (isAsync) this.raise(key.start, "Constructor can't be an async method")
+    method.kind = "constructor"
+  } else if (method.static && key.type === "Identifier" && key.name === "prototype") {
+    this.raise(key.start, "Classes may not have a static property named prototype")
```

Comment:
This could theoretically be moved into `parseClass`, since a field called `prototype` is not allowed, either. There's probably not gonna be a common way to get the name of a method or field (relevant discussion: estree/estree#154)